taichi/program/sparse_matrix.cpp

```diff
@@ -10,41 +10,86 @@ namespace lang {
 
 SparseMatrixBuilder::SparseMatrixBuilder(int rows,
                                          int cols,
-                                         int max_num_triplets)
-    : rows_(rows), cols_(cols), max_num_triplets_(max_num_triplets) {
-  data_.reserve(max_num_triplets * 3);
-  data_base_ptr_ = get_data_base_ptr();
+                                         int max_num_triplets,
+                                         DataType dtype)
+    : rows_(rows),
+      cols_(cols),
+      max_num_triplets_(max_num_triplets),
+      dtype_(dtype) {
+  auto element_size = data_type_size(dtype);
+  TI_ASSERT((element_size == 4 || element_size == 8));
+  data_base_ptr_ = new uchar[max_num_triplets_ * 3 * element_size];
+  if (data_base_ptr_ == nullptr) {
+    TI_ERROR("Failed to allocate memory for sparse matrix builder");
+  }
+}
+
+SparseMatrixBuilder::~SparseMatrixBuilder() {
+  // TODO: why this genenerates an segment fault error?
```

Comment:
I suggest debugging this out. We should not introduce memory problems into the codebase. BTW could you try to use `unique_ptr` instead of raw pointers?